src/lib/storage/index/adaptive_radix_tree/adaptive_radix_tree_index.cpp

```diff
@@ -22,8 +22,10 @@ size_t AdaptiveRadixTreeIndex::estimate_memory_consumption(ChunkOffset row_count
 }
 
 AdaptiveRadixTreeIndex::AdaptiveRadixTreeIndex(const std::vector<std::shared_ptr<const BaseSegment>>& segments_to_index)
-    : BaseIndex{get_index_type_of<AdaptiveRadixTreeIndex>()},
-      _indexed_segment(std::dynamic_pointer_cast<const BaseDictionarySegment>(segments_to_index.front())) {
+    : AbstractIndex{get_index_type_of<AdaptiveRadixTreeIndex>()},
+      _indexed_segment(segments_to_index.empty()
```

Comment:
When would this be the case? (other indexes, too)

Reply:
I think an empty segments_to_index vector wouldn't be passed in practice. But since the constructor is part of the (public) interface, indexes should either be able to handle an empty vector or throw an exception that it's not supported.

Reply:
Sounds like it should be an assertion. As this is our own interface, we could always remove it if there was a good reason to pass an empty vector. Until then, let's assume that whoever called it with an empty vector will be happy about being called out.

Reply:
But currently we even have the assertion a few lines below:
```c++
GroupKeyIndex::GroupKeyIndex(const std::vector<std::shared_ptr<const BaseSegment>>& segments_to_index)
    : AbstractIndex{get_index_type_of<GroupKeyIndex>()},
      _indexed_segment(segments_to_index.empty()
                           ? nullptr
                           : std::dynamic_pointer_cast<const BaseDictionarySegment>(segments_to_index[0])) {
  Assert(static_cast<bool>(_indexed_segment), "GroupKeyIndex only works with dictionary segments_to_index.");
  Assert((segments_to_index.size() == 1), "GroupKeyIndex only works with a single segment.");
```

`Assert(static_cast<bool>(_indexed_segment), "GroupKeyIndex only works with dictionary segments_to_index.")` does it. Would you rather use an assertion in the initializer list? If this is the case, how is it done?

Reply:
Ah, now I see the point of the ternary expression. Maybe add `// Illegal, but range check needed for accessing the first segment` after the `? nullptr`?

Reply:
agreed